File: src/src-v1.9-dnet/GKMapBrowser.pas
unit GKMapBrowser; {trans:fin}

{$I GEDKeeper2.inc}

interface

uses
  System.Drawing, System.ComponentModel, System.Windows.Forms, System.Resources,
  System.IO, System.Xml, System.Net, System.Text,
  VCLStub, GKUtils, GKEngine, GKCommon, GKMain;

type
  TMapBrowser = class({$IFNDEF DELPHI_NET}TWebBrowser{$ELSE}System.Windows.Forms.Control{$ENDIF})
  public
  type
    // класс точки на карте
    TGMapPoint = class(System.Object)
    private
      FDate: DateTime;
      FLatitude: System.Double;
      FLongitude: System.Double;
      FHint: string;
    public
      constructor Create(); overload;
      constructor Create(aLatitude, aLongitude: Double; aHint: string = ''); overload;

      property Date: DateTime read FDate write FDate;
      property Latitude: System.Double read FLatitude write FLatitude;    // широта
      property Longitude: System.Double read FLongitude write FLongitude; // долгота
      property Hint: string read FHint write FHint;                // подсказка
    end;
  strict private
  type
    // "прямоугольник" координат
    TCoordsRect = record
      MinLon, MinLat, MaxLon, MaxLat: Double;
    end;

  const
    GoogleKey = 'ABQIAAAAIcIQgkzLQ27NamNDh2wULxTh9o9-e_HqfKVqUrQPniGEP9J6uhSJmXGEipvip6lxpu_ZXrXaeHwWgQ';

    MapContent: string =
      '<html>'#13#10+
      '  <head>'#13#10+
      '    <meta http-equiv="content-type" content="text/html; charset=utf-8"/>'#13#10+
      '    <title>Map</title>'#13#10+
      '    <script src="http://maps.google.ru/maps?file=api&amp;v=2&amp;key='+GoogleKey+'" type="text/javascript"></script>'#13#10+
      '    <script type="text/javascript">'#13#10+
      '    var map;'#13#10+
      '    function addMarker(latitude, longitude, hint)'#13#10+
      '    {'#13#10+
      '        var blueIcon = new GIcon(G_DEFAULT_ICON);'#13#10+
      '        markerOptions = { icon : blueIcon, title : hint };'#13#10+
      '        var latlng = new GLatLng(latitude,longitude);'#13#10+
      '        map.addOverlay(new GMarker(latlng, markerOptions));'#13#10+
      '    }'#13#10+
      '    function initialize() {'#13#10+
      '      if (GBrowserIsCompatible()) {'#13#10+
      '        map = new GMap2(document.getElementById("map"));'#13#10+
      '        map.addMapType(G_PHYSICAL_MAP);'#13#10+
      '        map.setCenter(new GLatLng(55.755786, 37.617633), 11, G_PHYSICAL_MAP);'#13#10+
      '        map.addControl(new GLargeMapControl());'#13#10+
  //    '        map.addControl(new GSmallMapControl());'#13#10+
      '        map.addControl(new GMapTypeControl());'#13#10+
  //    '        map.addControl(new GOverviewMapControl());'#13#10+
  //    '        addMarker(59.944265, 30.319948, "This be standard marker without letter");'#13#10+
      '      }'#13#10+
      '    }'#13#10+
      '    </script>'#13#10+
      '  </head>'#13#10+
      '  <body onload="initialize()" onunload="GUnload()">'#13#10+
      '    <div id="map" style="position:absolute; width: 100%; height: 100%; left: 0px; top: 0px;"></div>'#13#10+
      '    <noscript><b style="font-family:Tahoma;">JavaScript must be switched on for use Google Maps.</b></noscript>'#13#10+
      '  </body>'#13#10+
      '</html>';

    var
      FMapFile: string;
      FMapPoints: TObjectList;
      FShowPoints: Boolean;
      FShowLines: Boolean;
      FUpdateCount: Integer;

    class var
      xmlDocument: System.Xml.XmlDocument;

    procedure gm_ClearPoints();
    procedure gm_ExecScript(Script: string);

    class function FillNode(Node: XMLNode): string;

    function GetMapPoint(Index: Integer): TGMapPoint;
    function GetMapPointsCount: Integer;
    function GetPointsFrame(): TCoordsRect;
    procedure SetVisibleElementes(Index: Integer; Value: Boolean);
  public
    constructor Create;
    destructor Destroy; override;

    function  AddPoint(aLatitude, aLongitude: Double; aHint: string = ''): Integer;
    procedure BeginUpdate();
    procedure ClearPoints();
    procedure DeletePoint(aIndex: Integer);
    procedure EndUpdate();
    procedure InitMap();
    procedure RefreshPoints();
    procedure SaveSnapshot(const aFileName: string);
    procedure SetCenter(aLatitude, aLongitude: Double; aScale: Integer = -1);
    procedure ZoomToBounds();

    property MapPoints[Index: Integer]: TGMapPoint read GetMapPoint; default;
    property MapPointsCount: Integer read GetMapPointsCount;
    property ShowPoints: Boolean index 0 read FShowPoints write SetVisibleElementes default True;
    property ShowLines: Boolean index 1 read FShowLines write SetVisibleElementes default False;

    class procedure GeoInit();
    class procedure GeoDone();
    class function GetInetFile(const FileURL: string; var Stream: System.IO.Stream): Boolean;
    class procedure RequestGeoCoords(SearchString: string; aPoints: TObjectList);
    class function RequestGeoAddress(aLatitude, aLongitude: Double): string;
  end;

implementation

{ TGMapPoint }

constructor TMapBrowser.TGMapPoint.Create();
begin
  inherited Create;
end;

constructor TMapBrowser.TGMapPoint.Create(aLatitude, aLongitude: Double; aHint: string = '');
begin
  inherited Create;

  FLatitude := aLatitude;
  FLongitude := aLongitude;
  FHint := aHint;
end;

{ TMapBrowser }

constructor TMapBrowser.Create;
begin
  inherited Create;

  FMapPoints := TObjectList.Create(True);
  FUpdateCount := 0;
  FShowPoints := True;
  FShowLines := True;
end;

destructor TMapBrowser.Destroy;
begin
  System.IO.File.Delete(FMapFile);

  ClearPoints();
  FMapPoints.Free;

  inherited Destroy;
end;

procedure TMapBrowser.InitMap();
var
  tf: System.IO.StreamWriter;
begin
  FMapFile := TGKUtils.GetTempDir() + 'GEDKeeperMap.html';

  tf := System.IO.StreamWriter.Create(FMapFile, False, Encoding.UTF8);
  tf.WriteLine(MapContent);
  tf.Close();

  {$IFNDEF DELPHI_NET}
  Navigate(FMapFile);
  {$ENDIF}
end;

function TMapBrowser.GetMapPointsCount(): Integer;
begin
  Result := FMapPoints.Count;
end;

function TMapBrowser.GetMapPoint(Index: Integer): TGMapPoint;
begin
  Result := nil;
  if (Index >= 0) and (Index < FMapPoints.Count)
  then Result := TGMapPoint(FMapPoints[Index]);
end;

procedure TMapBrowser.SetVisibleElementes(Index: Integer; Value: Boolean);
begin
  case Index of
    0: FShowPoints := Value;
    1: FShowLines := Value;
  end;
  RefreshPoints();
end;

procedure TMapBrowser.gm_ExecScript(Script: string);
begin
  if (Script.Trim() = '') then Exit;

  try
    {$IFNDEF DELPHI_NET}
    if (Document <> nil)
    then IHTMLDocument2(Document).parentWindow.execScript(Script, 'JavaScript');
    {$ENDIF}
  except
    on E: Exception do TGKUtils.LogWrite('TMapBrowser.gm_ExecScript(): '+E.Message);
  end;
end;

procedure TMapBrowser.gm_ClearPoints();
begin
  gm_ExecScript('map.clearOverlays();');
end;

function TMapBrowser.AddPoint(aLatitude, aLongitude: Double; aHint: string = ''): Integer;
var
  pt: TGMapPoint;
begin
  pt := TGMapPoint.Create(aLatitude, aLongitude, aHint);
  Result := FMapPoints.Add(pt);
end;

procedure TMapBrowser.DeletePoint(aIndex: Integer);
begin
  FMapPoints.Delete(aIndex);
  RefreshPoints();
end;

procedure TMapBrowser.BeginUpdate();
begin
  Inc(FUpdateCount);
end;

procedure TMapBrowser.EndUpdate();
begin
  Dec(FUpdateCount);
  if (FUpdateCount <= 0) then begin
    RefreshPoints();
    FUpdateCount := 0;
  end;
end;

procedure TMapBrowser.RefreshPoints();
var
  PointsScript, PolylineScript: string;
  i: Integer;
  pt: TGMapPoint;
begin
  // очистка точек только на карте, из списка не убиваются
  gm_ClearPoints();

  if MapPointsCount > 0 then begin
    for i := 0 to MapPointsCount - 1 do begin
      pt := MapPoints[i];
      // формируем скрипт создания точек
      PointsScript := PointsScript +
        System.&String.Format('addMarker({0:0.000000}, {1:0.000000}, "{2}");', [pt.Latitude, pt.Longitude, pt.Hint]);
      // формируем скрипт создания линий маршрута
      PolylineScript := PolylineScript + 'new GLatLng(' +
        pt.Latitude.ToString() + ',' + pt.Longitude.ToString() + '),';
    end;

    // если поднят флаг рисовать точки, то рисуем
    if ShowPoints
    then gm_ExecScript(PointsScript);

    // если поднят флаг рисовать линии
    if ShowLines then begin
      // добиваем скрипт
      StrDelete(PolylineScript, Length(PolylineScript), 1);
      PolylineScript := 'var polyline = new GPolyline([' +
        PolylineScript + '],"#FF0000",3);map.addOverlay(polyline);';
      // выполняем
      gm_ExecScript(PolylineScript);
    end;
  end;
end;

procedure TMapBrowser.SetCenter(aLatitude, aLongitude: Double; aScale: Integer = -1);
var
  Script: string;
begin
  if (aScale >= 0) then begin
    // если при вызове указан масштаб
    Script := 'var point = new GLatLng(' + aLatitude.ToString() + ',' + aLongitude.ToString() + '); ' +
      'map.setCenter(point, ' + aScale.ToString() + ')';
  end else begin
    // если не указан, то масштаб сохраняется
    Script := 'var point = new GLatLng(' + aLatitude.ToString() + ',' + aLongitude.ToString() + '); ' +
      'map.setCenter(point)';
  end;
  gm_ExecScript(Script);
end;

procedure TMapBrowser.ClearPoints();
begin
  gm_ClearPoints();
  FMapPoints.Clear;
end;

procedure TMapBrowser.ZoomToBounds();
var
  Script: string;
  Center_Latitude, Center_Longtude: Double;
  rt: TCoordsRect;
begin
  rt := GetPointsFrame();

  if (rt.MinLon <> rt.MaxLon) and (rt.MinLat <> rt.MaxLat) then begin
    Center_Longtude := (rt.MaxLon + rt.MinLon) / 2;
    Center_Latitude := (rt.MaxLat + rt.MinLat) / 2;

    Script := 'var point1 = new GLatLng({0:0.0000000}, {1:0.0000000});'+
              'var point2 = new GLatLng({2:0.0000000}, {3:0.0000000});'+
              'var bounds = new GLatLngBounds(point1, point2);'+
              'var zoom = map.getBoundsZoomLevel(bounds);'+
              'map.setCenter(new GLatLng({4:0.0000000}, {5:0.0000000}), zoom);';
    Script := System.&String.Format(Script, [rt.MinLat, rt.MinLon,
       rt.MaxLat, rt.MaxLon, Center_Latitude, Center_Longtude]);
    gm_ExecScript(Script);
  end;
end;

function TMapBrowser.GetPointsFrame(): TCoordsRect;
var
  i: Integer;
  pt: TGMapPoint;
begin
  if (FMapPoints.Count) > 0 then begin
    pt := TGMapPoint(FMapPoints[0]);

    Result.MinLon := pt.Longitude;
    Result.MaxLon := pt.Longitude;

    Result.MinLat := pt.Latitude;
    Result.MaxLat := pt.Latitude;

    if (FMapPoints.Count = 1) then begin
      Result.MinLon := Result.MinLon - 20;
      Result.MaxLon := Result.MaxLon + 20;

      Result.MinLat := Result.MinLat - 20;
      Result.MaxLat := Result.MaxLat + 20;

      Exit;
    end;

    for i := 0 to FMapPoints.Count - 1 do begin
      pt := TGMapPoint(FMapPoints[i]);

      if (Result.MinLon > pt.Longitude) then Result.MinLon := pt.Longitude
      else
      if (Result.MaxLon < pt.Longitude) then Result.MaxLon := pt.Longitude;

      if (Result.MinLat > pt.Latitude) then Result.MinLat := pt.Latitude
      else
      if (Result.MaxLat < pt.Latitude) then Result.MaxLat := pt.Latitude;
    end;
  end;
end;

procedure TMapBrowser.SaveSnapshot(const aFileName: string);
begin
  //
end;

class procedure TMapBrowser.GeoInit();
begin
  xmlDocument := System.Xml.XmlDocument.Create();
end;

class procedure TMapBrowser.GeoDone();
begin
  xmlDocument.Free; // interfaced object
end;

// отправка запроса и получение ответа
class function TMapBrowser.GetInetFile(const FileURL: string; var Stream: System.IO.Stream): Boolean;
var
  request: HttpWebRequest;
  response: HttpWebResponse;
  web_proxy: IWebProxy;
  proxy: TProxy;
begin
  try
    request := HttpWebRequest(WebRequest.CreateDefault(Uri.Create(FileURL)));
    request.ContentType := 'application/x-www-form-urlencoded';

    proxy := fmGEDKeeper.Options.Proxy;
    if (proxy.UseProxy) then begin
      web_proxy := WebProxy.Create(proxy.Server+':'+proxy.Port, True);
      //web_proxy.Credentials := NetworkCredential.Create(proxy.Login, proxy.Password);
      web_proxy.Credentials := CredentialCache.DefaultCredentials;
      request.Proxy := web_proxy;
    end;

    // Get Default Proxy
    (*IWebProxy web_proxy := WebRequest.GetSystemWebProxy();
    if (web_proxy <> nil) and (web_proxy.Credentials <> nil)
    then request.Proxy := web_proxy
    else begin
      web_proxy := WebProxy.Create(proxy.Server+':'+proxy.Port {'myproxy:8080'});
      web_proxy.Credentials := NetworkCredential.Create(proxy.Login, proxy.Password);
      request.Proxy := web_proxy;
    end;*)

    response := HttpWebResponse(request.GetResponse());
    Stream := response.GetResponseStream();
    Result := True;
  except
    on E: Exception do begin
      TGKUtils.LogWrite('GetInetFile(): ' + E.Message);
      Stream := nil;
      Result := False;
    end;
  end;
end;

// получение списка точек по запросу
class procedure TMapBrowser.RequestGeoCoords(SearchString: string; aPoints: TObjectList);
var
  Point: TGMapPoint;
  FileOnNet, coords: string;
  Stream: System.IO.Stream;
  //strd: System.IO.StreamReader;
  //st: string;
  Node, PlacemarkNode, PointNode, AddressNode: XMLNode;
  i: Integer;
  coord_tokens: array of string;
begin
  if (aPoints = nil) then Exit;

  //StringList := TStringList.Create;
  try
    SearchString := StringReplace(SearchString.Trim(), ' ', '+', [rfReplaceAll]);
    FileOnNet := 'http://maps.google.ru/maps/geo?q={0}&output=xml&key={1}&gl=ru';
    FileOnNet := System.&String.Format(FileOnNet, [SearchString, GoogleKey]);
    if GetInetFile(FileOnNet, Stream) then begin
      xmlDocument.Load(Stream);

      {strd := StreamReader.Create(Stream);
      st := strd.ReadToEnd;
      TGKUtils.Hole(st);}

      Node := xmlDocument.DocumentElement;
      Node := Node.Item['Response'];
      if (Node <> nil) and (Node.ChildNodes.Count > 0) then
        for i := 0 to Node.ChildNodes.Count - 1 do
          if (Node.ChildNodes[i].Name = 'Placemark') then begin
            // находим узел точки
            PlacemarkNode := Node.ChildNodes[i];
            // получаем узел адреса
            AddressNode := PlacemarkNode.Item['address'];
            //получаем узел координат
            PointNode := PlacemarkNode.Item['Point'];
            PointNode := PointNode.Item['coordinates'];
            if (AddressNode <> nil) and (PointNode <> nil) then begin
              Point := TGMapPoint.Create;
              // получаем адрес
              Point.Hint := AddressNode.InnerText;
              // получаем координаты
              coords := PointNode.InnerText;
              // разбираем координаты
              coord_tokens := coords.Split([',']);
              if (Length(coord_tokens) > 1) then begin
                // Формируем точку
                Point.Longitude := TGKUtils.StrToFloatDef(coord_tokens[0], -1);
                Point.Latitude := TGKUtils.StrToFloatDef(coord_tokens[1], -1);
                // добавляем точку в список
                if (Point.Latitude <> -1) and (Point.Longitude <> -1)
                then aPoints.Add(Point);
              end else Point.Free;
            end;
          end;
    end;
  finally
    Stream.Free;
  end;
end;

class function TMapBrowser.FillNode(Node: XMLNode): string;
var
  i: Integer;
begin
  if (Node.ChildNodes <> nil) and (Node.ChildNodes.Count > 0) then
    for i := 0 to Node.ChildNodes.Count - 1 do begin
      FillNode(Node.ChildNodes[i]);
      // самый первый адрес наиболее подробный, берем его
      if (Node.Name = 'address') then begin
        Result := Node.InnerText;
        Break;
      end;
    end;
end;

// запрос обратного геокодирования
class function TMapBrowser.RequestGeoAddress(aLatitude, aLongitude: Double): string;
var
  FileOnNet: string;
  Stream: System.IO.Stream;
begin
  Result := '';
  try
    FileOnNet := 'http://maps.google.ru/maps/geo?ll={0:0.000000},{1:0.000000}&output=xml&key={2}&gl=ru';
    FileOnNet := System.&String.Format(FileOnNet, [aLatitude, aLongitude, GoogleKey]);
    if GetInetFile(FileOnNet, Stream) then begin
      xmlDocument.Load(Stream);
      Result := FillNode(xmlDocument.DocumentElement);
    end;
  finally
    Stream.Free;
  end;
end;

end.

File: src/GKUI/Controls/GKMapBrowser.cs
using System;
using System.Globalization;
using System.IO;
using System.Net;
using System.Text;
using System.Windows.Forms;
using System.Xml;

using Ext.Utils;
using GKCore;

/// <summary>
/// Localization: clean
/// </summary>

namespace GKUI.Controls
{
	public class GKMapBrowser : WebBrowser, IDisposable
	{
		public class TGMapPoint
		{
			private DateTime FDate;
			private double FLatitude;
			private double FLongitude;
			private string FHint;

			public DateTime Date
			{
				get { return this.FDate; }
				set { this.FDate = value; }
			}

			public double Latitude
			{
				get { return this.FLatitude; }
				set { this.FLatitude = value; }
			}

			public double Longitude
			{
				get { return this.FLongitude; }
				set { this.FLongitude = value; }
			}

			public string Hint
			{
				get { return this.FHint; }
				set { this.FHint = value; }
			}

			public TGMapPoint()
			{
			}

			public TGMapPoint(double aLatitude, double aLongitude, string aHint)
			{
				this.FLatitude = aLatitude;
				this.FLongitude = aLongitude;
				this.FHint = aHint;
			}

			public void Free()
			{
				SysUtils.Free(this);
			}
		}


		private struct TCoordsRect
		{
			public double MinLon;
			public double MinLat;
			public double MaxLon;
			public double MaxLat;
		}

		private static readonly string MapContent;
		private string FMapFile;
		private TList FMapPoints;
		private bool FShowPoints;
		private bool FShowLines;
		private int FUpdateCount;
		private static XmlDocument xmlDocument;

		public bool ShowPoints
		{
			get { return this.FShowPoints; }
			set { this.SetVisibleElementes(0, value); }
		}

		public bool ShowLines
		{
			get { return this.FShowLines; }
			set { this.SetVisibleElementes(1, value); }
		}

		private void gm_ClearPoints()
		{
			this.gm_ExecScript("map.clearOverlays();");
		}

		private void gm_ExecScript(string Script)
		{
			if (Script.Trim() != "")
			{
				try
				{
					mshtml.IHTMLWindow2 win = (mshtml.IHTMLWindow2)this.Document.Window.DomWindow;
					win.execScript(Script, "JavaScript");
				}
				catch (Exception E)
				{
					SysUtils.LogWrite("TMapBrowser.gm_ExecScript(): " + E.Message);
				}
			}
		}

		private static string FillNode(XmlNode Node)
		{
			string Result = "";
			if (Node.ChildNodes != null && Node.ChildNodes.Count > 0)
			{
				int num = Node.ChildNodes.Count - 1;
				int i = 0;
				if (num >= i)
				{
					num++;
					while (true)
					{
						GKMapBrowser.FillNode(Node.ChildNodes[i]);
						if (Node.Name == "address") {
							break;
						}
						i++;
						if (i == num) {
							return Result;
						}
					}
					Result = Node.InnerText;
				}
			}
			return Result;
		}

		public GKMapBrowser.TGMapPoint GetMapPoint(int Index)
		{
			GKMapBrowser.TGMapPoint Result = null;
			if (Index >= 0 && Index < this.FMapPoints.Count)
			{
				Result = (this.FMapPoints[Index] as GKMapBrowser.TGMapPoint);
			}
			return Result;
		}

		public int GetMapPointsCount()
		{
			return this.FMapPoints.Count;
		}

		private GKMapBrowser.TCoordsRect GetPointsFrame()
		{
			GKMapBrowser.TCoordsRect Result = new TCoordsRect();
			if (this.FMapPoints.Count > 0)
			{
				GKMapBrowser.TGMapPoint pt = this.FMapPoints[0] as GKMapBrowser.TGMapPoint;
				Result.MinLon = pt.Longitude;
				Result.MaxLon = pt.Longitude;
				Result.MinLat = pt.Latitude;
				Result.MaxLat = pt.Latitude;
				if (this.FMapPoints.Count == 1)
				{
					Result.MinLon = (Result.MinLon - 20.0);
					Result.MaxLon = (Result.MaxLon + 20.0);
					Result.MinLat = (Result.MinLat - 20.0);
					Result.MaxLat = (Result.MaxLat + 20.0);
				}
				else
				{
					int num = this.FMapPoints.Count - 1;
					for (int i = 0; i <= num; i++)
					{
						pt = (this.FMapPoints[i] as GKMapBrowser.TGMapPoint);

						if (Result.MinLon > pt.Longitude) Result.MinLon = pt.Longitude;
						else if (Result.MaxLon < pt.Longitude) Result.MaxLon = pt.Longitude;

						if (Result.MinLat > pt.Latitude) Result.MinLat = pt.Latitude;
						else if (Result.MaxLat < pt.Latitude) Result.MaxLat = pt.Latitude;
					}
				}
			}
			return Result;
		}

		private void SetVisibleElementes(int Index, bool Value)
		{
			if (Index != 0)
			{
				if (Index == 1)
				{
					this.FShowLines = Value;
				}
			}
			else
			{
				this.FShowPoints = Value;
			}
			this.RefreshPoints();
		}

		public GKMapBrowser()
		{
			this.FMapPoints = new TList(true);
			this.FUpdateCount = 0;
			this.FShowPoints = true;
			this.FShowLines = true;
		}

		protected override void Dispose(bool Disposing)
		{
			if (Disposing)
			{
				File.Delete(this.FMapFile);
				this.ClearPoints();
				this.FMapPoints.Dispose();
			}
			base.Dispose(Disposing);
		}

		public int AddPoint(double aLatitude, double aLongitude, string aHint)
		{
			GKMapBrowser.TGMapPoint pt = new GKMapBrowser.TGMapPoint(aLatitude, aLongitude, aHint);
			return this.FMapPoints.Add(pt);
		}

		public void BeginUpdate()
		{
			this.FUpdateCount++;
		}

		public void ClearPoints()
		{
			this.gm_ClearPoints();
			this.FMapPoints.Clear();
		}

		public void DeletePoint(int aIndex)
		{
			this.FMapPoints.Delete(aIndex);
			this.RefreshPoints();
		}

		public void EndUpdate()
		{
			this.FUpdateCount--;
			if (this.FUpdateCount <= 0)
			{
				this.RefreshPoints();
				this.FUpdateCount = 0;
			}
		}

		public void InitMap()
		{
			this.FMapFile = TGenEngine.GetTempDir() + "\\GEDKeeperMap.html";
			StreamWriter tf = new StreamWriter(this.FMapFile, false, Encoding.UTF8);
			tf.WriteLine(GKMapBrowser.MapContent);
			tf.Close();

			Navigate(FMapFile);
		}

		public static string CoordToStr(double val) {
			NumberFormatInfo nfi = new NumberFormatInfo();
			nfi.NumberDecimalSeparator = ".";
			return val.ToString("0.000000", nfi);
		}

		public void RefreshPoints()
		{
			this.gm_ClearPoints();
			if (this.FMapPoints.Count > 0)
			{
				string PointsScript = "";
				string PolylineScript = "";

				for (int i = 0; i <= this.FMapPoints.Count - 1; i++)
				{
					TGMapPoint pt = this.FMapPoints[i] as TGMapPoint;
					PointsScript += string.Format("addMarker({0}, {1}, \"{2}\");", new object[]
					{ CoordToStr(pt.Latitude), CoordToStr(pt.Longitude), pt.Hint });

					PolylineScript = string.Concat(new string[]
					{
						PolylineScript, "new GLatLng(",	
						CoordToStr(pt.Latitude), ",", CoordToStr(pt.Longitude), "),"
					});
				}

				if (this.ShowPoints)
				{
					this.gm_ExecScript(PointsScript);
				}

				if (this.ShowLines)
				{
					int num2 = (PolylineScript != null) ? PolylineScript.Length : 0;
					PolylineScript = PolylineScript.Remove(num2 - 1, 1);
					PolylineScript = "var polyline = new GPolyline([" + PolylineScript + "],\"#FF0000\",3);map.addOverlay(polyline);";
					this.gm_ExecScript(PolylineScript);
				}
			}
		}

		public void SaveSnapshot(string aFileName)
		{
		}

		public void SetCenter(double aLatitude, double aLongitude, int aScale)
		{
			string Script;
			if (aScale >= 0) {
				Script = string.Concat(new string[] {
					"var point = new GLatLng(", CoordToStr(aLatitude), ",", CoordToStr(aLongitude), "); ", 
					"map.setCenter(point, ", aScale.ToString(), ")"
				});
			} else {
				Script = string.Concat(new string[] {
					"var point = new GLatLng(", CoordToStr(aLatitude), ",", CoordToStr(aLongitude), "); ", 
					"map.setCenter(point)"
				});
			}

			this.gm_ExecScript(Script);
		}

		public void ZoomToBounds()
		{
			GKMapBrowser.TCoordsRect rt = this.GetPointsFrame();
			if (rt.MinLon != rt.MaxLon && rt.MinLat != rt.MaxLat)
			{
				double Center_Longtude = ((rt.MaxLon + rt.MinLon) / 2.0);
				double Center_Latitude = ((rt.MaxLat + rt.MinLat) / 2.0);

				string Script = "var point1 = new GLatLng({0}, {1});var point2 = new GLatLng({2}, {3});var bounds = new GLatLngBounds(point1, point2);var zoom = map.getBoundsZoomLevel(bounds);map.setCenter(new GLatLng({4}, {5}), zoom);";
				Script = string.Format(Script, new object[]
				{ CoordToStr(rt.MinLat), CoordToStr(rt.MinLon), CoordToStr(rt.MaxLat), CoordToStr(rt.MaxLon), CoordToStr(Center_Latitude), CoordToStr(Center_Longtude) });

				this.gm_ExecScript(Script);
			}
		}

		public static void GeoInit()
		{
			GKMapBrowser.xmlDocument = new XmlDocument();
		}

		public static void GeoDone()
		{
			GKMapBrowser.xmlDocument = null;
		}

		public static bool GetInetFile(string FileURL, ref Stream Stream)
		{
			bool Result;
			try
			{
				HttpWebRequest request = WebRequest.CreateDefault(new Uri(FileURL)) as HttpWebRequest;
				request.ContentType = "application/x-www-form-urlencoded";
				ProxyOptions proxy = GKUI.TfmGEDKeeper.Instance.Options.Proxy;
				if (proxy.UseProxy)
				{
					request.Proxy = new WebProxy(proxy.Server + ":" + proxy.Port, true)
					{
						Credentials = CredentialCache.DefaultCredentials
					};
				}
				HttpWebResponse response = request.GetResponse() as HttpWebResponse;
				Stream = response.GetResponseStream();
				Result = true;
			}
			catch (Exception E)
			{
				SysUtils.LogWrite("GetInetFile(): " + E.Message);
				Stream = null;
				Result = false;
			}
			return Result;
		}

		public static void RequestGeoCoords(string SearchString, TList aPoints)
		{
			if (aPoints != null)
			{
				Stream stm = null;
				try
				{
					SearchString = SearchString.Trim().Replace(" ", "+");

					string FileOnNet = "http://maps.google.ru/maps/geo?q={0}&output=xml&key={1}&gl=ru";
					FileOnNet = string.Format(FileOnNet, new object[]
					{
						SearchString, 
						"ABQIAAAAIcIQgkzLQ27NamNDh2wULxTh9o9-e_HqfKVqUrQPniGEP9J6uhSJmXGEipvip6lxpu_ZXrXaeHwWgQ"
					});
					if (GKMapBrowser.GetInetFile(FileOnNet, ref stm))
					{
						GKMapBrowser.xmlDocument.Load(stm);
						XmlNode Node = GKMapBrowser.xmlDocument.DocumentElement;
						Node = Node["Response"];
						if (Node != null && Node.ChildNodes.Count > 0)
						{
							int num = Node.ChildNodes.Count - 1;
							for (int i = 0; i <= num; i++)
							{
								if (Node.ChildNodes[i].Name == "Placemark")
								{
									XmlNode PlacemarkNode = Node.ChildNodes[i];
									XmlNode AddressNode = PlacemarkNode["address"];
									XmlNode PointNode = PlacemarkNode["Point"];
									PointNode = PointNode["coordinates"];
									if (AddressNode != null && PointNode != null)
									{
										GKMapBrowser.TGMapPoint pt = new GKMapBrowser.TGMapPoint();
										pt.Hint = AddressNode.InnerText;
										string coords = PointNode.InnerText;
										string[] coord_tokens = coords.Split(new char[] {','});
										if (((coord_tokens != null) ? coord_tokens.Length : 0) > 1)
										{
											pt.Longitude = SysUtils.ParseFloat(coord_tokens[0], -1.0);
											pt.Latitude = SysUtils.ParseFloat(coord_tokens[1], -1.0);
											if (pt.Latitude != -1.0 && pt.Longitude != -1.0) aPoints.Add(pt);
										}
										else
										{
											pt.Free();
										}
									}
								}
							}
						}
					}
				}
				finally
				{
					//stm.Dispose; checkit
				}
			}
		}

		public static string RequestGeoAddress(double aLatitude, double aLongitude)
		{
			string Result = "";
			Stream stm = null;
			try
			{
				string FileOnNet = "http://maps.google.ru/maps/geo?ll={0},{1}&output=xml&key={2}&gl=ru";
				FileOnNet = string.Format(FileOnNet, new object[]
				{
				    CoordToStr(aLatitude), CoordToStr(aLongitude),
					"ABQIAAAAIcIQgkzLQ27NamNDh2wULxTh9o9-e_HqfKVqUrQPniGEP9J6uhSJmXGEipvip6lxpu_ZXrXaeHwWgQ"
				});
				if (GKMapBrowser.GetInetFile(FileOnNet, ref stm))
				{
					GKMapBrowser.xmlDocument.Load(stm);
					Result = GKMapBrowser.FillNode(GKMapBrowser.xmlDocument.DocumentElement);
				}
			}
			finally
			{
				//stm.Dispose; //checkit
			}
			return Result;
		}

		static GKMapBrowser()
		{
			GKMapBrowser.MapContent = "<html>\r\n  <head>\r\n    <meta http-equiv=\"content-type\" content=\"text/html; charset=utf-8\"/>\r\n    <title>Map</title>\r\n    <script src=\"http://maps.google.ru/maps?file=api&amp;v=2&amp;key=ABQIAAAAIcIQgkzLQ27NamNDh2wULxTh9o9-e_HqfKVqUrQPniGEP9J6uhSJmXGEipvip6lxpu_ZXrXaeHwWgQ\" type=\"text/javascript\"></script>\r\n    <script type=\"text/javascript\">\r\n    var map;\r\n    function addMarker(latitude, longitude, hint)\r\n    {\r\n        var blueIcon = new GIcon(G_DEFAULT_ICON);\r\n        markerOptions = { icon : blueIcon, title : hint };\r\n        var latlng = new GLatLng(latitude,longitude);\r\n        map.addOverlay(new GMarker(latlng, markerOptions));\r\n    }\r\n    function initialize() {\r\n      if (GBrowserIsCompatible()) {\r\n        map = new GMap2(document.getElementById(\"map\"));\r\n        map.addMapType(G_PHYSICAL_MAP);\r\n        map.setCenter(new GLatLng(55.755786, 37.617633), 11, G_PHYSICAL_MAP);\r\n        map.addControl(new GLargeMapControl());\r\n        map.addControl(new GMapTypeControl());\r\n      }\r\n    }\r\n    </script>\r\n  </head>\r\n  <body onload=\"initialize()\" onunload=\"GUnload()\">\r\n    <div id=\"map\" style=\"position:absolute; width: 100%; height: 100%; left: 0px; top: 0px;\"></div>\r\n    <noscript><b style=\"font-family:Tahoma;\">JavaScript must be switched on for use Google Maps.</b></noscript>\r\n  </body>\r\n</html>";
		}
	}
}
